### PR TITLE
Short term fix for the EIA parser

### PR DESCRIPTION
**Pull request by unitrium**
## Issue

We are currently seeing frequent instances of the EIA parser returning incomplete data points due to the EIA publishing incomplete data. This has consequences on our estimation models as they are picking up the trend on missing modes.

We are working on our outlier detection layer but it's not ready yet to handle those cases. Therefore as a short term fix we would like the parser to discard incomplete datapoints so that our estimation models are not poisoned by those.

## Description

Create a general filter static method for production breakdown using the zone's capacity. If a zone has capacity for a mode we expect it to have a value for this mode at all time which is not None.

### Preview

Testing in progress.

### Double check

- [ ] I have tested my parser changes locally with `poetry run test_parser "zone_key"`
- [ ] I have run `pnpx prettier --write .` and `poetry run format` to format my changes.

### CODE REVIEW

**Review by mathilde-daugy — APPROVED** [review]
looks good, the code is very clear. Thank you for addressing this so quickly :)

**Review by VIKTORVAV99 — CHANGES_REQUESTED** [review]
A small fix is needed.

Also we might want to consider relaxing the storage modes, I don't think for example US-TEX-ERCO has any at all reported.
> ➜  electricitymaps-contrib git:(robin/ele-3593-short-term-fix-for-expected-modes-on-the-eia-parser) ✗ poetry run test_parser "US-TEX-ERCO"
WARNING:test_parser:Discarded production event for US-TEX-ERCO at 2024-01-11 06:00:00+00:00 due to missing battery storage value.
WARNING:test_parser:Discarded production event for US-TEX-ERCO at 2024-01-11 07:00:00+00:00 due to missing battery storage value.
WARNING:test_parser:Discarded production event for US-TEX-ERCO at 2024-01-11 08:00:00+00:00 due to missing battery storage value.
WARNING:test_parser:Discarded production event for US-TEX-ERCO at 2024-01-11 09:00:00+00:00 due to missing battery storage value.
WARNING:test_parser:Discarded production event for US-TEX-ERCO at 2024-01-11 10:00:00+00:00 due to missing battery storage value.
WARNING:test_parser:Discarded production event for US-TEX-ERCO at 2024-01-11 11:00:00+00:00 due to missing battery storage value.
WARNING:test_parser:Discarded production event for US-TEX-ERCO at 2024-01-11 12:00:00+00:00 due to missing battery storage value.
WARNING:test_parser:Discarded production event for US-TEX-ERCO at 2024-01-11 13:00:00+00:00 due to missing battery storage value.
WARNING:test_parser:Discarded production event for US-TEX-ERCO at 2024-01-11 14:00:00+00:00 due to missing battery storage value.
WARNING:test_parser:Discarded production event for US-TEX-ERCO at 2024-01-11 15:00:00+00:00 due to missing battery storage value.
WARNING:test_parser:Discarded production event for US-TEX-ERCO at 2024-01-11 16:00:00+00:00 due to missing battery storage value.
WARNING:test_parser:Discarded production event for US-TEX-ERCO at 2024-01-11 17:00:00+00:00 due to missing battery storage value.
WARNING:test_parser:Discarded production event for US-TEX-ERCO at 2024-01-11 18:00:00+00:00 due to missing battery storage value.
WARNING:test_parser:Discarded production event for US-TEX-ERCO at 2024-01-11 19:00:00+00:00 due to missing battery storage value.
WARNING:test_parser:Discarded production event for US-TEX-ERCO at 2024-01-11 20:00:00+00:00 due to missing battery storage value.
WARNING:test_parser:Discarded production event for US-TEX-ERCO at 2024-01-11 21:00:00+00:00 due to missing battery storage value.
WARNING:test_parser:Discarded production event for US-TEX-ERCO at 2024-01-11 22:00:00+00:00 due to missing battery storage value.
WARNING:test_parser:Discarded production event for US-TEX-ERCO at 2024-01-11 23:00:00+00:00 due to missing battery storage value.
WARNING:test_parser:Discarded production event for US-TEX-ERCO at 2024-01-12 00:00:00+00:00 due to missing battery storage value.
WARNING:test_parser:Discarded production event for US-TEX-ERCO at 2024-01-12 01:00:00+00:00 due to missing battery storage value.
WARNING:test_parser:Discarded production event for US-TEX-ERCO at 2024-01-12 02:00:00+00:00 due to missing battery storage value.
WARNING:test_parser:Discarded production event for US-TEX-ERCO at 2024-01-12 03:00:00+00:00 due to missing battery storage value.
WARNING:test_parser:Discarded production event for US-TEX-ERCO at 2024-01-12 04:00:00+00:00 due to missing battery storage value.
WARNING:test_parser:Discarded production event for US-TEX-ERCO at 2024-01-12 05:00:00+00:00 due to missing battery storage value.
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/viktor/Repos/electricitymaps-contrib/.venv/lib/python3.10/site-packages/click/core.py", line 1134, in __call__
    return self.main(*args, **kwargs)
  File "/Users/viktor/Repos/electricitymaps-contrib/.venv/lib/python3.10/site-packages/click/core.py", line 1059, in main
    rv = self.invoke(ctx)
  File "/Users/viktor/Repos/electricitymaps-contrib/.venv/lib/python3.10/site-packages/click/core.py", line 1401, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/viktor/Repos/electricitymaps-contrib/.venv/lib/python3.10/site-packages/click/core.py", line 767, in invoke
    return __callback(*args, **kwargs)
  File "/Users/viktor/Repos/electricitymaps-contrib/test_parser.py", line 74, in test_parser
    raise ValueError(f"Error: parser returned nothing ({res})")
ValueError: Error: parser returned nothing ([])
➜  electricitymaps-contrib git:(robin/ele-3593-short-term-fix-for-expected-modes-on-the-eia-parser) ✗

File: electricitymap/contrib/lib/models/event_lists.py [hunk 1]
```diff
@@ -245,6 +247,37 @@ def merge_production_breakdowns(
...
+                mode for mode, capacity_value in capacity.items() if capacity_value > 0
+            ]
+            for mode in required_modes:
+                value = event.get_value(mode)
+                if value is None:
+                    valid = False
+                    events.logger.warning(
+                        f"Discarded production event for {event.zone_key} at {event.datetime} due to missing {mode} value."
+                    )
+                    break
```
[the remark this answers is not in the document]
Comment (unitrium) [reply electricitymap/contrib/lib/models/event_lists.py:271]: This is a bit of a dumb suggestion looks like it doesn't understand the code 😅

**Comment by mathilde-daugy** [comment]
> e might want to consider relaxing the storage modes, I d

Good point, it's likely that the storage modes are by default None and not 0

**Comment by unitrium** [comment]
Hum there's an issue there's no biomass currently in the EIA parser but we have biomass capacity:
https://github.com/electricitymaps/electricitymaps-contrib/blob/611f1ff36fb78c71b3c1c7e5b806038d98b9b7a4/parsers/EIA.py#L342-L343

**Comment by VIKTORVAV99** [comment]
Looks like it's working as expected now, but there are a few zones that seems to have capacity for a few modes that are not reported at all. For example it seems like Oil is not reported for US-TEX-ERCO (really picked the worst(best) one to test with 😅). Not sure if we should handle that in some way on a per zone key level?

**Comment by unitrium** [comment]
True, I'm thinking yes we could activate this for a few zones where we saw the issue happen recently and add others as they go.

**Comment by unitrium** [comment]
@VIKTORVAV99 I added a zone selection to configure it only for certain zones and specific keys to ignore.

**Review by VIKTORVAV99 — APPROVED** [review]
Nice! Works as expected now!

**Pull request merged**